File: basic/java/tree/binaryTree/SameTree.java
package java.tree.binaryTree;


// Given the roots of two binary trees p and q, write a function to check if they are the same or not.
// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.



public class SameTree {
    
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        
        TreeNode() {}
        
        TreeNode(int val) {
            this.val = val;
        }
        
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null || p.val != q.val) return false;
        // Recursively check the left subtree and the right subtree
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}




// // usage:
// public class Main {
//     public static void main(String[] args) {
//         // Create first pair of trees that are the same
//         TreeNode node1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
//         TreeNode node2 = new TreeNode(1, new TreeNode(2), new TreeNode(3));

//         // Create second pair of trees that are different
//         TreeNode node3 = new TreeNode(1, new TreeNode(2), null);
//         TreeNode node4 = new TreeNode(1, null, new TreeNode(2));

//         SameTree sol = new SameTree();
        
//         // Check if first pair of trees are the same
//         System.out.println("Tree 1 and Tree 2 are the same: " + sol.isSameTree(node1, node2)); // true
        
//         // Check if second pair of trees are the same
//         System.out.println("Tree 3 and Tree 4 are the same: " + sol.isSameTree(node3, node4)); // false
//     }
// }

File: basic/python/tree/binaryTree/sameTree.py
from typing import Optional


# Given the roots of two binary trees p and q, write a function to check if they are the same or not.
# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # If both nodes are None, then we've reached the end of both trees at this branch
        if not p and not q:
            return True
        # If only one of the nodes is None or the values do not match, trees are not the same
        if not p or not q or p.val != q.val:
            return False
        # Recursively check the left and right subtrees
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)



# Usage:
node1 = TreeNode(1, TreeNode(2), TreeNode(3))
node2 = TreeNode(1, TreeNode(2), TreeNode(3))
sol = Solution()
print(sol.isSameTree(node1, node2))  # True

node3 = TreeNode(1, TreeNode(2))
node4 = TreeNode(1, None, TreeNode(2))
print(sol.isSameTree(node3, node4))  # False
